prepare/cli.py

```diff
@@ -0,0 +1,160 @@
+import click
+
+from numpy import unique, ndarray
+from scipy.sparse.csc import csc_matrix
+from os.path import isfile, isdir, splitext, expanduser
+
+settings = dict(help_option_names=['-h', '--help'])
+
+
+@click.command()
+@click.argument('dataset', nargs=1, metavar='<dataset: file or path to data>', required=True)
+@click.option('--layout', default='umap', type=click.Choice(['umap', 'tsne', 'umap+tsne']),
+              help='layout algorithm', show_default=True)
+@click.option('--recipe', default='none', type=click.Choice(['none', 'seurat', 'zheng17']),
+              help='preprocessing to run', show_default=True)
+@click.option('--output', default='', help='save a new file to filename')
+@click.option('--set-obs-names', default='', help='named field to set as index for obs')
+@click.option('--set-var-names', default='', help='named field to set as index for var')
+@click.option('--make-obs-names-unique', default=True, is_flag=True, help='ensure obs index is unique', show_default=True)
+@click.option('--make-var-names-unique', default=True, is_flag=True, help='ensure var index is unique', show_default=True)
+@click.option('--sparse', default=False, is_flag=True, help='whether to force sparsity', show_default=True)
+@click.option('--overwriting', default=False, is_flag=True, help='allow file overwriting', show_default=True)
+@click.option('--plotting', default=False, is_flag=True, help='whether to generate plots', show_default=True)
+def cli(dataset, layout, recipe, output, set_obs_names, set_var_names,
+        make_obs_names_unique, make_var_names_unique, sparse, overwriting, plotting):
+    """
+    preprocesses data for use with cellxgene
+    """
+
+    # collect slow imports here to make CLI startup more responsive
+    click.echo('[cellxgene] Starting CLI...')
+    import matplotlib
+    matplotlib.use('Agg')
+    import scanpy.api as sc
+
+    # scanpy settings
+    sc.settings.verbosity = 0
+    sc.settings.autosave = True
+
+    # check args
+    if sparse and not recipe == 'none':
+        raise click.UsageError('Cannot use a recipe when forcing sparsity')
+
+    output = expanduser(output)
+    if isfile(output) and not overwrite:
+        raise click.UsageError('Cannot overwwrite existing file %s, try using the flag --overwrite' % output)
+
+    def load_data(dataset):
+        if isfile(dataset):
+            name, extension = splitext(dataset)
+            if extension == '.h5ad':
+                adata = sc.read_h5ad(dataset)
+            elif extension == '.loom':
+                adata = sc.read_loom(dataset)
+            else:
+                raise click.FileError(dataset, hint='does not have a valid extension [.h5ad | .loom]')
+        elif isdir(dataset):
+            if not dataset.endswith('/'):
+                raise click.FileError(dataset, hint='a path must end with a /')
```

Comment:
Why are we forcing this?

Reply:
will fix by automatically appending a separator